File: app/src/main/java/ch/bailu/aat/gpx/tools/SimplifierDistance.java
package ch.bailu.aat.gpx.tools;

import ch.bailu.aat.gpx.AltitudeDelta;
import ch.bailu.aat.gpx.AutoPause;
import ch.bailu.aat.gpx.GpxDeltaHelper;
import ch.bailu.aat.gpx.GpxList;
import ch.bailu.aat.gpx.GpxListWalker;
import ch.bailu.aat.gpx.GpxPoint;
import ch.bailu.aat.gpx.GpxPointFirstNode;
import ch.bailu.aat.gpx.GpxPointNode;
import ch.bailu.aat.gpx.GpxSegmentNode;
import ch.bailu.aat.gpx.MaxSpeed;

public class SimplifierDistance extends GpxListWalker {

    private final static float MIN_DISTANCE=25f;

    private GpxList newList;
    private boolean newSegment = true;

    private GpxPointNode lastPoint=null;


    @Override
    public boolean doList(GpxList track) {
        newList = new GpxList(track.getDelta().getType(),
                MaxSpeed.NULL,  AutoPause.NULL, AltitudeDelta.NULL);
        return true;
    }

    @Override
    public boolean doSegment(GpxSegmentNode segment) {
        newSegment = true;
        return true;
    }

    @Override
    public boolean doMarker(GpxSegmentNode marker) {
        return true;
    }

    @Override
    public void doPoint(GpxPointNode point) {
        if (newSegment) {
            newSegment = false;

            newList.appendToNewSegment(new GpxPoint(point), point.getAttributes());
            lastPoint = point;

        } else {

            if (isLastInSegment(point) || hasDistance(point)) {
                newList.appendToCurrentSegment(new GpxPoint(point), point.getAttributes());
                lastPoint = point;
            }

        }
    }


    private boolean hasDistance(GpxPointNode point) {
        return GpxDeltaHelper.getDistance(lastPoint, point) >= MIN_DISTANCE;
    }


    private boolean isLastInSegment(GpxPointNode point) {
        return point.getNext() == null || point.getNext() instanceof GpxPointFirstNode;
    }


    public GpxList getNewList() {
        return newList;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/tools/SimplifierDistance.kt
package ch.bailu.aat_lib.gpx.tools

import ch.bailu.aat_lib.gpx.GpxDeltaHelper
import ch.bailu.aat_lib.gpx.GpxList
import ch.bailu.aat_lib.gpx.GpxListWalker
import ch.bailu.aat_lib.gpx.GpxPoint
import ch.bailu.aat_lib.gpx.GpxPointFirstNode
import ch.bailu.aat_lib.gpx.GpxPointNode
import ch.bailu.aat_lib.gpx.GpxSegmentNode
import ch.bailu.aat_lib.gpx.attributes.GpxListAttributes

class SimplifierDistance : GpxListWalker() {
    var newList: GpxList? = null
        private set
    private var newSegment = true
    private var lastPoint: GpxPointNode? = null

    override fun doList(track: GpxList): Boolean {
        newList = GpxList(track.getDelta().getType(), GpxListAttributes.NULL)
        return true
    }

    override fun doSegment(segment: GpxSegmentNode): Boolean {
        newSegment = true
        return true
    }

    override fun doMarker(marker: GpxSegmentNode): Boolean {
        return true
    }

    override fun doPoint(point: GpxPointNode) {
        if (newSegment) {
            newSegment = false
            newList?.appendToNewSegment(GpxPoint(point), point.getAttributes())
            lastPoint = point
        } else {
            if (isLastInSegment(point) || hasDistance(point)) {
                newList?.appendToCurrentSegment(GpxPoint(point), point.getAttributes())
                lastPoint = point
            }
        }
    }

    private fun hasDistance(point: GpxPointNode): Boolean {
        return GpxDeltaHelper.getDistance(lastPoint, point) >= MIN_DISTANCE
    }

    private fun isLastInSegment(point: GpxPointNode): Boolean {
        return point.next == null || point.next is GpxPointFirstNode
    }

    companion object {
        private const val MIN_DISTANCE = 25f
    }
}
